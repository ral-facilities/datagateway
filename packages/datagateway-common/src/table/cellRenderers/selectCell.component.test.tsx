import React from 'react';
import { shallow } from 'enzyme';
import SelectCell from './selectCell.component';

describe('Select cell component', () => {
  const setLastChecked = jest.fn();
  const onCheck = jest.fn();
  const onUncheck = jest.fn();
  const data = [
    {
      id: 1,
      name: 'test 1',
    },
    {
      id: 2,
      name: 'test 2',
    },
    {
      id: 3,
      name: 'test 3',
    },
  ];
  const selectCellProps = {
    columnIndex: 1,
    dataKey: 'test',
    isScrolling: false,
    rowIndex: 2,
    rowData: data[2],
    selectedRows: [],
    data,
    lastChecked: -1,
    loading: false,
    setLastChecked,
    onCheck,
    onUncheck,
    loading: false,
  };

  afterEach(() => {
    setLastChecked.mockClear();
    onCheck.mockClear();
    onUncheck.mockClear();
  });

  it('renders correctly when unchecked', () => {
    const wrapper = shallow(<SelectCell {...selectCellProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly when checked', () => {
    const wrapper = shallow(
      <SelectCell {...selectCellProps} selectedRows={[3]} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly when selectedRows is undefined', () => {
    const wrapper = shallow(
      <SelectCell {...selectCellProps} selectedRows={undefined} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly when selectedRows loading is true', () => {
    const wrapper = shallow(
      <SelectCell
        {...selectCellProps}
        loading={true}
        selectedRows={undefined}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls setLastChecked when checkbox is clicked', () => {
    const wrapper = shallow(<SelectCell {...selectCellProps} />);

    wrapper.find('.tour-dataview-add-to-cart').prop('onClick')(
      new MouseEvent('click')
    );
    expect(setLastChecked).toHaveBeenCalledWith(2);
  });

  it('calls onCheck when the row is unselected and the checkbox is clicked', () => {
    const wrapper = shallow(<SelectCell {...selectCellProps} />);

    wrapper.find('.tour-dataview-add-to-cart').prop('onClick')(
      new MouseEvent('click')
    );
    expect(onCheck).toHaveBeenCalledWith([3]);
  });

  it('calls onUncheck when the row is selected and the checkbox is clicked', () => {
    const wrapper = shallow(
      <SelectCell {...selectCellProps} selectedRows={[3]} />
    );

    wrapper.find('.tour-dataview-add-to-cart').prop('onClick')(
      new MouseEvent('click')
    );
    expect(onUncheck).toHaveBeenCalledWith([3]);
  });

  it('calls onCheck when the row is selected via shift-click and the checkbox is clicked', () => {
    const wrapper = shallow(
      <SelectCell {...selectCellProps} lastChecked={0} />
    );

    wrapper.find('.tour-dataview-add-to-cart').prop('onClick')(
      new MouseEvent('click', { shiftKey: true })
    );
    expect(onCheck).toHaveBeenCalledWith([1, 2, 3]);
  });

  it('calls onUncheck when the row is unselected via shift-click and the checkbox is clicked', () => {
    const wrapper = shallow(
      <SelectCell
        {...selectCellProps}
        lastChecked={0}
        selectedRows={[1, 2, 3]}
      />
    );

    wrapper.find('.tour-dataview-add-to-cart').prop('onClick')(
      new MouseEvent('click', { shiftKey: true })
    );
    expect(onUncheck).toHaveBeenCalledWith([1, 2, 3]);
  });
});
