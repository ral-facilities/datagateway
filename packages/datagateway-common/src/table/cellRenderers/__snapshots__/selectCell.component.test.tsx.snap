// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select cell component renders correctly when checked 1`] = `
<ForwardRef(TableCell)
  component="div"
  padding="checkbox"
  size="small"
  variant="body"
>
  <StyledTooltip
    placement="right"
    title=""
  >
    <span>
      <ForwardRef(Checkbox)
        checked={true}
        checkedIcon={
          <Memo
            fontSize="small"
          />
        }
        className="tour-dataview-add-to-cart"
        disabled={false}
        icon={
          <Memo
            fontSize="small"
          />
        }
        inputProps={
          Object {
            "aria-label": "select row 2",
          }
        }
        onClick={[Function]}
        size="small"
        style={
          Object {
            "padding": "inherit",
          }
        }
      />
    </span>
  </StyledTooltip>
</ForwardRef(TableCell)>
`;

exports[`Select cell component renders correctly when selectedRows is undefined 1`] = `
<ForwardRef(TableCell)
  component="div"
  padding="checkbox"
  size="small"
  variant="body"
>
  <StyledTooltip
    placement="right"
    title="buttons.cart_loading_failed_tooltip"
  >
    <span>
      <ForwardRef(Checkbox)
        checkedIcon={
          <Memo
            fontSize="small"
          />
        }
        className="tour-dataview-add-to-cart"
        disabled={true}
        icon={
          <Memo
            fontSize="small"
          />
        }
        inputProps={
          Object {
            "aria-label": "select row 2",
          }
        }
        onClick={[Function]}
        size="small"
        style={
          Object {
            "padding": "inherit",
          }
        }
      />
    </span>
  </StyledTooltip>
</ForwardRef(TableCell)>
`;

exports[`Select cell component renders correctly when selectedRows loading is true 1`] = `
<ForwardRef(TableCell)
  component="div"
  padding="checkbox"
  size="small"
  variant="body"
>
  <StyledTooltip
    placement="right"
    title="buttons.cart_loading_tooltip"
  >
    <span>
      <ForwardRef(Checkbox)
        checkedIcon={
          <Memo
            fontSize="small"
          />
        }
        className="tour-dataview-add-to-cart"
        disabled={true}
        icon={
          <Memo
            fontSize="small"
          />
        }
        inputProps={
          Object {
            "aria-label": "select row 2",
          }
        }
        onClick={[Function]}
        size="small"
        style={
          Object {
            "padding": "inherit",
          }
        }
      />
    </span>
  </StyledTooltip>
</ForwardRef(TableCell)>
`;

exports[`Select cell component renders correctly when unchecked 1`] = `
<ForwardRef(TableCell)
  component="div"
  padding="checkbox"
  size="small"
  variant="body"
>
  <StyledTooltip
    placement="right"
    title=""
  >
    <span>
      <ForwardRef(Checkbox)
        checked={false}
        checkedIcon={
          <Memo
            fontSize="small"
          />
        }
        className="tour-dataview-add-to-cart"
        disabled={false}
        icon={
          <Memo
            fontSize="small"
          />
        }
        inputProps={
          Object {
            "aria-label": "select row 2",
          }
        }
        onClick={[Function]}
        size="small"
        style={
          Object {
            "padding": "inherit",
          }
        }
      />
    </span>
  </StyledTooltip>
</ForwardRef(TableCell)>
`;
